python/lsst/daf/persistence/butler.py

```diff
@@ -85,63 +97,46 @@ class Butler(object):
-            else:
-                raise RuntimeError(
-                        "No mapper provided and no %s available" %
-                        (mapperFile,))
-        mapperFile = os.path.join(basePath, mapperFile)
-
-        # Read the name of the mapper class and instantiate it
-        with open(mapperFile, "r") as f:
-            mapperName = f.readline().strip()
-        components = mapperName.split(".")
-        if len(components) <= 1:
-            raise RuntimeError("Unqualified mapper name %s in %s" %
-                    (mapperName, mapperFile))
-        pkg = importlib.import_module(".".join(components[:-1]))
-        return getattr(pkg, components[-1])
+    @classmethod
+    def cfg(cls, repoCfg):
+        return Policy({'repoCfg':repoCfg})
 
-    def __init__(self, root, mapper=None, **mapperArgs):
-        """Construct the Butler.  If no mapper class is provided, then a file
-        named "_mapper" is expected to be found in the repository, which
-        must be a filesystem path.  The first line in that file is read and
-        must contain the fully-qualified name of a Mapper subclass, which is
-        then imported and instantiated using the root and the mapperArgs.
-
-        @param root (str)       the repository to be managed (at least
-                                initially).  May be None if a mapper is
-                                provided.
-        @param mapper (Mapper)  if present, the Mapper subclass instance
-                                to be used as the butler's mapper.
-        @param **mapperArgs     arguments to be passed to the mapper's
-                                __init__ method, in addition to the root.
-        """
 
+    def initWithCfg(self, cfg):
+        self._cfg = cfg
         self.datasetTypeAliasDict = {}
 
-        if mapper is not None:
-            self.mapper = mapper
-        else:
-            cls = Butler.getMapperClass(root)
-            self.mapper = cls(root=root, **mapperArgs)
+        self.repository = repository.Repository.Repository(cfg['repoCfg'])
 
         # Always use an empty Persistence policy until we can get rid of it
         persistencePolicy = pexPolicy.Policy()
         self.persistence = Persistence.getPersistence(persistencePolicy)
-        self.log = pexLog.Log(pexLog.Log.getDefaultLog(),
-                "daf.persistence.butler")
+        self.log = pexLog.Log(pexLog.Log.getDefaultLog(), "daf.persistence.butler")
+
+
+    def __init__(self, root, mapper=None, **mapperArgs):
+        # prefered init arg is to pass a single arg; an instance of butlerCfg:
+        # butler = Butler(butlerCfg(...))
+        # for backward compatibility: init signature can take a posix root path, and optionally a mapper class
+        # that will be init'ed with mapperArgs.
+        if (isinstance(root, Policy)):
+            config = root
+        else:
+            parentCfg = posixRepoCfg(root=root, mapper=mapper, mapperArgs=mapperArgs)
```

Comment:
put a comment here:
it's not preferred to have a repo be both input (parent) and output (self/peer), and will likely break any provenance system we have in place. but it allows backward compatibility.